cfg/githubfieldmapper.go

```diff
@@ -0,0 +1,235 @@
+package cfg
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/andygrunwald/go-jira"
+	"github.com/coreos/issue-sync/lib/models"
+	"strings"
+	"time"
+)
+
+type FieldMapper interface {
+	MapFields(issue *models.ExtendedGithubIssue) jira.IssueFields
+
+	GetFieldValue(jIssue *jira.Issue, fieldKey FieldKey) (interface{}, error)
+
+	// getFieldIDs requests the metadata of every issue field in the JIRA
+	// project, and saves the IDs of the custom fields used by issue-sync.
+	GetFieldIDs(client jira.Client) (map[FieldKey]string, error)
+}
+
+type DefaultFieldMapper struct {
+	Config *Config
+}
+
+type JsonFieldMapper struct {
+	Config *Config
+}
+
+func (m DefaultFieldMapper) GetFieldValue(jIssue *jira.Issue, fieldKey FieldKey) (interface{}, error) {
+	switch fieldKey {
+	case GitHubID:
+		fallthrough
+	case GitHubNumber:
+		return jIssue.Fields.Unknowns.Int(m.Config.GetCompleteFieldKey(fieldKey))
+	default:
+		result, exists := jIssue.Fields.Unknowns.Value(m.Config.GetCompleteFieldKey(fieldKey))
+		if !exists {
+			return nil, errors.New("Field not found")
+		}
+		return result, nil
+	}
+}
+
+func (m DefaultFieldMapper) MapFields(issue *models.ExtendedGithubIssue) jira.IssueFields {
+	fields := jira.IssueFields{
+		Type: jira.IssueType{
+			Name: "Task", // TODO: Determine issue type
+		},
+		Project:     m.Config.GetProject(),
+		Summary:     issue.GetTitle(),
+		Description: issue.GetBody(),
+		Labels: 	 issue.CommitIds,
+		Unknowns:    map[string]interface{}{},
+	}
+
+	fields.Unknowns[m.Config.GetCompleteFieldKey(GitHubID)] = issue.GetID()
```

Comment:
Here's another example of duplicated code that could be eliminated. Maybe create a function where you would pass in the gh field and the jira field, and it would stuff the value into the right place in fields.Unknowns

Reply:
Not sure I agree with this. All that would be factored out is the field access. I'd need to call this function just as many times as there are key/value pairs. And there would be no other re-use outside of this function. Putting Unknowns into a variable would achieve the same since its a reference but I think that is bad style.

Reply:
Seeing a string of about 40 characters appear verbatim in several lines of code makes me cringe. It's not about the number of calls, or optimizing for speed; it's about clarity and maintenance. I agree that putting Unknowns in a variable wouldn't help. If you're not persuaded, we should discuss in person so we can really understand each other.